File: src/main/java/mediathek/controller/history/AboHistoryController.java
package mediathek.controller.history;

import ca.odell.glazedlists.BasicEventList;
import ca.odell.glazedlists.EventList;
import mediathek.config.StandardLocations;
import mediathek.gui.messages.history.AboHistoryChangedEvent;
import mediathek.tool.MessageBus;
import okhttp3.HttpUrl;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.jetbrains.annotations.NotNull;

import java.io.*;
import java.nio.channels.FileChannel;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.StandardOpenOption;
import java.util.ArrayList;
import java.util.List;
import java.util.Set;
import java.util.concurrent.ConcurrentHashMap;

public class AboHistoryController {
    private static final String FILENAME = "downloadAbos.txt";
    private static final Logger logger = LogManager.getLogger();
    /**
     * Quick lookup set for history checks.
     * Stores only URLs.
     */
    private final Set<String> urlCacheSet = ConcurrentHashMap.newKeySet();
    /**
     * The actual storage for all history data.
     * Will be written to file.
     * MUST BE LOCKED DURING MANIPULATION!
     */
    private final EventList<MVUsedUrl> baseDataEventList = new BasicEventList<>();
    private Path urlPath;

    public AboHistoryController() {
        final var settingsDir = StandardLocations.getSettingsDirectory();
        try {
            urlPath = settingsDir.resolve(FILENAME);
            if (Files.notExists(urlPath))
                Files.createFile(urlPath);
        } catch (IOException e) {
            logger.error("I/O error occured", e);
            urlPath = null;
        }

        listeBauen();
    }

    public EventList<MVUsedUrl> getDataList() {
        return baseDataEventList;
    }

    private void sendChangeMessage() {
        MessageBus.getMessageBus().publishAsync(new AboHistoryChangedEvent());
    }

    private void clearLists() {
        try {
            baseDataEventList.getReadWriteLock().writeLock().lock();
            baseDataEventList.clear();
            urlCacheSet.clear();
        } finally {
            baseDataEventList.getReadWriteLock().writeLock().unlock();
        }
    }

    /**
     * Remove all stored entries.
     * Also deletes the used text file. When supported it will be moved to trash, otherwise deleted.
     */
    public void removeAll() {
        clearLists();

        try (var fc = FileChannel.open(urlPath, StandardOpenOption.WRITE)){
            fc.truncate(0);
        } catch (IOException e) {
            logger.error("Could not empty abo history file", e);
        }

        sendChangeMessage();
    }

    /**
     * Check if a URL exists in the cache set.
     *
     * @param urlFilm a string url
     * @return true if exists in cache, false otherwise.
     */
    public boolean urlExists(@NotNull String urlFilm) {
        return urlCacheSet.contains(urlFilm);
    }

    public synchronized void removeUrl(@NotNull String urlFilm) {
        //Logfile einlesen, entsprechende Zeile Filtern und dann Logfile überschreiben
        //wenn die URL im Logfile ist, dann true zurück
        boolean entryFound = false;

        // if the url is NOT in our list, it won´t be in the file...bail out
        if (!urlExists(urlFilm))
            return;

        final List<String> liste = new ArrayList<>();
        try (InputStream is = Files.newInputStream(urlPath);
             InputStreamReader isr = new InputStreamReader(is);
             LineNumberReader in = new LineNumberReader(isr)) {
            String zeile;
            while ((zeile = in.readLine()) != null) {
                if (MVUsedUrl.getUrlAusZeile(zeile).getUrl().equals(urlFilm)) {
                    entryFound = true; //nur dann muss das Logfile auch geschrieben werden
                } else {
                    liste.add(zeile);
                }
            }
        } catch (Exception ex) {
            logger.error("urlAusLogfileLoeschen(String)", ex);
        }

        //und jetzt wieder schreiben, wenn nötig
        if (entryFound) {
            try (OutputStream os = Files.newOutputStream(urlPath);
                 OutputStreamWriter osw = new OutputStreamWriter(os);
                 BufferedWriter bufferedWriter = new BufferedWriter(osw)) {
                for (String entry : liste)
                    bufferedWriter.write(entry + '\n');
            } catch (Exception ex) {
                logger.error("urlAusLogfileLoeschen(String)", ex);
            }

            clearLists();

            listeBauen();

            sendChangeMessage();
        }
    }

    public void add(@NotNull MVUsedUrl usedUrl) {
        try {
            baseDataEventList.getReadWriteLock().writeLock().lock();
            baseDataEventList.add(usedUrl);
            urlCacheSet.add(usedUrl.getUrl());
        } finally {
            baseDataEventList.getReadWriteLock().writeLock().unlock();
        }

        try (OutputStream os = Files.newOutputStream(urlPath, StandardOpenOption.APPEND);
             OutputStreamWriter osw = new OutputStreamWriter(os);
             BufferedWriter bufferedWriter = new BufferedWriter(osw)) {
            bufferedWriter.write(usedUrl.getPreparedRowString());
        } catch (Exception ex) {
            logger.error("Single add failed", ex);
        }

        sendChangeMessage();
    }

    /**
     * Create the internally used list from text file.
     */
    private void listeBauen() {
        List<String> badEntriesList = new ArrayList<>();

        try (InputStream is = Files.newInputStream(urlPath);
             InputStreamReader isr = new InputStreamReader(is);
             LineNumberReader in = new LineNumberReader(isr)) {
            String zeile;

            baseDataEventList.getReadWriteLock().writeLock().lock();

            while ((zeile = in.readLine()) != null) {
                MVUsedUrl mvuu = MVUsedUrl.getUrlAusZeile(zeile);
                var url = mvuu.getUrl();
                if (url.startsWith("rtmp:")) {
                    //logger.warn("RTMP URL found in file {}, skipping: {}", urlPath, url);
                    badEntriesList.add(zeile);
                    continue;
                }

                var okHttpUrl = HttpUrl.parse(url);
                if (okHttpUrl == null) {
                    //logger.warn("Invalid URL received in {}, skipping: {}", urlPath,url);
                    badEntriesList.add(zeile);
                    continue;
                }

                // so far so good, add to lists
                baseDataEventList.add(mvuu);
                urlCacheSet.add(url);
            }
        } catch (Exception ex) {
            logger.error("listeBauen()", ex);
        } finally {
            baseDataEventList.getReadWriteLock().writeLock().unlock();
        }

        checkBadEntries(badEntriesList);

        printDebugInfo();
    }

    private void printDebugInfo() {
        try {
            baseDataEventList.getReadWriteLock().readLock().lock();
            logger.trace("urlCacheSet size: {} for file {}", urlCacheSet.size(), urlPath);
            logger.trace("dataList size: {} for file {}", baseDataEventList.size(), urlPath);
        } finally {
            baseDataEventList.getReadWriteLock().readLock().unlock();
        }
    }

    private void checkBadEntries(@NotNull List<String> badEntriesList) {
        if (!badEntriesList.isEmpty()) {
            logger.warn("File {} contained {} invalid entries ", urlPath, badEntriesList.size());
            removeIllegalEntries(badEntriesList);
            badEntriesList.clear();
        }
    }

    /**
     * Append multiple urls to history file.
     *
     * @param mvuuList the items to add.
     */
    public void add(@NotNull List<MVUsedUrl> mvuuList) {
        try (OutputStream os = Files.newOutputStream(urlPath, StandardOpenOption.APPEND);
             OutputStreamWriter osw = new OutputStreamWriter(os);
             BufferedWriter bufferedWriter = new BufferedWriter(osw)) {
            baseDataEventList.getReadWriteLock().writeLock().lock();

            for (MVUsedUrl mvuu : mvuuList) {
                baseDataEventList.add(mvuu);
                urlCacheSet.add(mvuu.getUrl());
                bufferedWriter.write(mvuu.getPreparedRowString());
            }
        } catch (Exception ex) {
            logger.error("zeilenSchreiben()", ex);
        } finally {
            baseDataEventList.getReadWriteLock().writeLock().unlock();
        }
        sendChangeMessage();
    }

    private synchronized void removeIllegalEntries(@NotNull List<String> badEntriesList) {
        logger.trace("Cleaning entries for {}", urlPath);

        final List<String> cleanedEntriesList = new ArrayList<>();
        try (InputStream is = Files.newInputStream(urlPath);
             InputStreamReader isr = new InputStreamReader(is);
             LineNumberReader in = new LineNumberReader(isr)) {
            String zeile;
            while ((zeile = in.readLine()) != null) {
                if (!badEntriesList.contains(zeile)) {
                    cleanedEntriesList.add(zeile);
                }
            }
        } catch (Exception ex) {
            logger.error("removeIllegalEntries()", ex);
        }

        try (OutputStream os = Files.newOutputStream(urlPath);
             OutputStreamWriter osw = new OutputStreamWriter(os);
             BufferedWriter bufferedWriter = new BufferedWriter(osw)) {
            for (var entry : cleanedEntriesList)
                bufferedWriter.write(entry + '\n');
        } catch (Exception ex) {
            logger.error("removeIllegalEntries()", ex);
        }

        cleanedEntriesList.clear();
        logger.trace("Finished cleaning entries for {}", urlPath);
    }
}

File: src/main/kotlin/mediathek/controller/history/AboHistoryController.kt
/*
 * Copyright (c) 2026 derreisende77.
 * This code was developed as part of the MediathekView project https://github.com/mediathekview/MediathekView
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package mediathek.controller.history

import kotlinx.coroutines.CoroutineDispatcher
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.runBlocking
import kotlinx.coroutines.withContext
import mediathek.config.StandardLocations
import mediathek.gui.messages.history.AboHistoryChangedEvent
import mediathek.tool.MessageBus
import org.apache.logging.log4j.LogManager
import org.sqlite.SQLiteConfig
import org.sqlite.SQLiteDataSource
import java.io.IOException
import java.net.URI
import java.nio.charset.StandardCharsets
import java.nio.file.Files
import java.nio.file.Path
import java.nio.file.StandardCopyOption
import java.sql.Connection
import java.sql.PreparedStatement
import java.sql.ResultSet
import java.sql.SQLException
import java.util.concurrent.ConcurrentHashMap

class AboHistoryController(
    private val legacyFilePath: Path = StandardLocations.getSettingsDirectory().resolve(LEGACY_FILENAME),
    databasePath: Path = StandardLocations.getSettingsDirectory().resolve(DATABASE_FILENAME),
    private val databaseDispatcher: CoroutineDispatcher = Dispatchers.IO.limitedParallelism(1)
) {
    private val dataSource = createDataSource(databasePath)
    private val urlCache = ConcurrentHashMap.newKeySet<String>()

    init {
        AboHistoryDatabaseBootstrapper(databasePath, dataSource).bootstrap()
        runOnDatabase {
            migrateLegacyFileIfPresent()
            reloadUrlCache()
        }
    }

    fun getDataList(): List<AboHistoryEntry> = runOnDatabase { loadEntries() }

    fun urlExists(urlFilm: String): Boolean = isSupportedUrl(urlFilm) && urlCache.contains(urlFilm)

    fun removeAll() {
        val changed = runOnDatabase { deleteAllEntries() }
        if (changed) {
            updateCacheAndNotify(
                cacheMutation = urlCache::clear
            )
        }
    }

    fun removeUrl(urlFilm: String) = removeUrls(listOf(urlFilm))

    fun removeUrls(urlsToRemove: Collection<String>): Int {
        val supportedUrls = urlsToRemove.normalizedSupportedUrls()
        if (supportedUrls.isEmpty()) {
            return 0
        }

        val removedCount = runOnDatabase { deleteEntries(supportedUrls) }
        if (removedCount > 0) {
            val removedUrlSet = supportedUrls.toSet()
            updateCacheAndNotify {
                urlCache.removeAll(removedUrlSet)
            }
        }
        return removedCount
    }

    fun add(entry: AboHistoryEntry) = add(listOf(entry))

    fun add(entries: List<AboHistoryEntry>) {
        val candidates = entries.filter(::isPersistable)
        if (candidates.isEmpty()) {
            return
        }

        val changed = runOnDatabase { insertEntries(candidates) }
        if (changed) {
            val addedUrls = candidates.asSequence().map { it.url }.toSet()
            updateCacheAndNotify {
                urlCache.addAll(addedUrls)
            }
        }
    }

    private fun migrateLegacyFileIfPresent() {
        if (Files.notExists(legacyFilePath)) {
            return
        }

        val importedEntries = readLegacyEntries() ?: return
        if (importedEntries.isNotEmpty()) {
            insertEntries(importedEntries)
        }
        archiveLegacyFile()
    }

    private fun readLegacyEntries(): List<AboHistoryEntry>? {
        return try {
            Files.newBufferedReader(legacyFilePath, StandardCharsets.UTF_8).use { reader ->
                generateSequence { reader.readLine() }
                    .map(String::trim)
                    .filter(String::isNotEmpty)
                    .mapNotNull(::parseLegacyLine)
                    .filter(::isPersistable)
                    .toList()
            }
        } catch (e: IOException) {
            logger.error("Could not import legacy abo history from {}", legacyFilePath, e)
            null
        }
    }

    private fun archiveLegacyFile() {
        val archivedPath = legacyFilePath.resolveSibling(legacyFilePath.fileName.toString() + ARCHIVED_SUFFIX)
        try {
            Files.move(legacyFilePath, archivedPath, StandardCopyOption.REPLACE_EXISTING)
        } catch (e: IOException) {
            logger.warn("Could not archive migrated abo history file {}", legacyFilePath, e)
        }
    }

    private fun createDataSource(dbPath: Path): SQLiteDataSource {
        val config = SQLiteConfig().apply {
            setEncoding(SQLiteConfig.Encoding.UTF8)
            setJournalMode(SQLiteConfig.JournalMode.WAL)
            setSynchronous(SQLiteConfig.SynchronousMode.NORMAL)
            setBusyTimeout(5_000)
            enforceForeignKeys(false)
            enableLoadExtension(false)
        }

        return SQLiteDataSource(config).also {
            it.url = "jdbc:sqlite:${dbPath.toAbsolutePath()}"
        }
    }

    private fun openConnection(): Connection =
        dataSource.connection.apply {
            transactionIsolation = Connection.TRANSACTION_SERIALIZABLE
        }

    private fun <T> runOnDatabase(block: suspend () -> T): T = runBlocking {
        withContext(databaseDispatcher) {
            block()
        }
    }

    private fun isPersistable(entry: AboHistoryEntry): Boolean = isSupportedUrl(entry.url)

    private fun isSupportedUrl(url: String): Boolean {
        if (url.isBlank()) {
            return false
        }
        if (url.startsWith("rtmp:", ignoreCase = true)) {
            return false
        }
        if (!(url.startsWith("http://", ignoreCase = true) || url.startsWith("https://", ignoreCase = true))) {
            return false
        }
        if (url.any(Char::isWhitespace)) {
            return false
        }

        return try {
            val uri = URI(url)
            uri.isAbsolute &&
                (uri.scheme.equals("http", ignoreCase = true) || uri.scheme.equals("https", ignoreCase = true)) &&
                !uri.host.isNullOrBlank()
        } catch (_: IllegalArgumentException) {
            false
        }
    }

    private fun loadEntries(): List<AboHistoryEntry> {
        return try {
            openConnection().use { connection ->
                connection.prepareStatement(SELECT_ALL_SQL).use { statement ->
                    statement.executeQuery().use { resultSet ->
                        buildList {
                            while (resultSet.next()) {
                                resultSet.toAboHistoryEntryOrNull()?.let(::add)
                            }
                        }
                    }
                }
            }
        } catch (e: SQLException) {
            logger.error("Could not load abo history entries", e)
            emptyList()
        }
    }

    private fun insertEntries(entries: List<AboHistoryEntry>): Boolean {
        return runLoggedDatabaseOperation(
            errorMessage = "Could not append abo history entries",
            fallback = false
        ) {
            openConnection().use { connection ->
                connection.inTransaction {
                    prepareStatement(INSERT_SQL).use { statement ->
                        entries.forEach { entry -> statement.bindInsertEntry(entry) }
                        countAffectedRows(statement.executeBatch()) > 0
                    }
                }
            }
        }
    }

    private fun deleteAllEntries(): Boolean {
        return runLoggedDatabaseOperation(
            errorMessage = "Could not update abo history",
            fallback = false
        ) {
            openConnection().use { connection ->
                connection.prepareStatement(DELETE_ALL_SQL).use { statement ->
                    statement.executeUpdate() > 0
                }
            }
        }
    }

    private fun deleteEntries(urlsToRemove: List<String>): Int {
        return runLoggedDatabaseOperation(
            errorMessage = "Could not delete abo history entries",
            fallback = 0
        ) {
            openConnection().use { connection ->
                connection.inTransaction {
                    prepareStatement(DELETE_SQL).use { statement ->
                        urlsToRemove.forEach { url -> statement.bindDeleteUrl(url) }
                        countAffectedRows(statement.executeBatch())
                    }
                }
            }
        }
    }

    private fun reloadUrlCache() {
        val loadedUrls = runLoggedDatabaseOperation(
            errorMessage = "Could not reload abo history cache",
            fallback = emptySet()
        ) {
            loadEntries().asSequence()
                .map(AboHistoryEntry::url)
                .toSet()
        }
        urlCache.clear()
        urlCache.addAll(loadedUrls)
    }

    private fun parseLegacyLine(line: String): AboHistoryEntry? {
        if (!line.contains(LEGACY_ENTRY_SEPARATOR)) {
            return null
        }

        val urlSeparatorIndex = line.lastIndexOf(LEGACY_ENTRY_SEPARATOR)
        if (urlSeparatorIndex <= 0) {
            return null
        }

        val metadata = line.substring(0, urlSeparatorIndex)
        val url = line.substring(urlSeparatorIndex + LEGACY_ENTRY_SEPARATOR.length).trim()

        val firstSeparatorIndex = metadata.indexOf(LEGACY_FIELD_SEPARATOR)
        if (firstSeparatorIndex < 0) {
            return null
        }

        val secondSeparatorIndex = metadata.indexOf(
            LEGACY_FIELD_SEPARATOR,
            firstSeparatorIndex + LEGACY_FIELD_SEPARATOR.length
        )
        if (secondSeparatorIndex < 0) {
            return null
        }

        val date = metadata.substring(0, firstSeparatorIndex).trim()
        val theme = metadata.substring(firstSeparatorIndex + LEGACY_FIELD_SEPARATOR.length, secondSeparatorIndex).trim()
        val title = metadata.substring(secondSeparatorIndex + LEGACY_FIELD_SEPARATOR.length).trim()

        return createEntryOrNull(date, theme, title, url)
    }

    private fun <T> Connection.inTransaction(block: Connection.() -> T): T {
        val previousAutoCommit = autoCommit
        autoCommit = false
        return try {
            val result = block()
            commit()
            result
        } catch (e: SQLException) {
            rollback()
            throw e
        } finally {
            autoCommit = previousAutoCommit
        }
    }

    private fun PreparedStatement.bindInsertEntry(entry: AboHistoryEntry) {
        setString(1, entry.formattedDate)
        setString(2, entry.theme)
        setString(3, entry.title)
        setString(4, entry.url)
        addBatch()
    }

    private fun PreparedStatement.bindDeleteUrl(url: String) {
        setString(1, url)
        addBatch()
    }

    private fun countAffectedRows(results: IntArray): Int =
        results.count { it > 0 || it == PreparedStatement.SUCCESS_NO_INFO }

    private fun Collection<String>.normalizedSupportedUrls(): List<String> =
        asSequence()
            .filter(::isSupportedUrl)
            .distinct()
            .toList()

    private fun updateCacheAndNotify(cacheMutation: () -> Unit) {
        cacheMutation()
        publishChangeEvent()
    }

    private fun <T> runLoggedDatabaseOperation(errorMessage: String, fallback: T, operation: () -> T): T {
        return try {
            operation()
        } catch (e: SQLException) {
            logger.error(errorMessage, e)
            fallback
        }
    }

    private fun ResultSet.toAboHistoryEntryOrNull(): AboHistoryEntry? =
        createEntryOrNull(
            getString("datum"),
            getString("thema"),
            getString("titel"),
            getString("url")
        )

    private fun createEntryOrNull(date: String, theme: String, title: String, url: String): AboHistoryEntry? {
        return AboHistoryEntry.parse(date, theme, title, url).also { entry ->
            if (entry == null) {
                logger.warn("Ignoring abo history entry with invalid date \"{}\" and URL \"{}\"", date, url)
            }
        }
    }

    private fun publishChangeEvent() {
        MessageBus.messageBus.publishAsync(AboHistoryChangedEvent())
    }

    private companion object {
        private val logger = LogManager.getLogger()
        private const val LEGACY_FILENAME = "downloadAbos.txt"
        private const val DATABASE_FILENAME = "abo-history.db"
        private const val ARCHIVED_SUFFIX = ".migrated"
        private const val LEGACY_ENTRY_SEPARATOR = "  |###|  "
        private const val LEGACY_FIELD_SEPARATOR = " |#| "
        private const val SELECT_ALL_SQL = """
            SELECT datum, thema, titel, url
            FROM abo_history
            ORDER BY id ASC
        """
        private const val INSERT_SQL = """
            INSERT OR IGNORE INTO abo_history(datum, thema, titel, url)
            VALUES (?, ?, ?, ?)
        """
        private const val DELETE_SQL = """
            DELETE FROM abo_history
            WHERE url = ?
        """
        private const val DELETE_ALL_SQL = "DELETE FROM abo_history"
    }
}
